File: src/main/java/dev/tuhkanens/comfortlib/implementation/ConfigImpl.java
package dev.tuhkanens.comfortlib.implementation;

import dev.tuhkanens.comfortlib.api.ConfigAPI;
import org.spongepowered.configurate.ConfigurateException;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.yaml.YamlConfigurationLoader;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;

public class ConfigImpl implements ConfigAPI {

    private ConfigurationNode root;
    private Path path;

    @Override
    public void create(Class<?> clazz, Path directory) {
        String fileName = "config.yml";

        path = directory.resolve(fileName);

        if (!Files.exists(path)) {
            try (InputStream input = clazz.getClassLoader().getResourceAsStream(fileName)) {
                if (input == null) {
                    throw new IllegalStateException("Default config.yml not found in resources!");
                }
                Files.createDirectories(directory);
                Files.copy(input, path);
            } catch (IOException e) {
                throw new IllegalStateException("Failed to copy config.yml", e);
            }
        }

        load();
    }

    @Override
    public void reload() {
        load();
    }

    @Override
    public ConfigurationNode getNode() {
        return root;
    }

    private void load() {
        YamlConfigurationLoader yaml = YamlConfigurationLoader.builder()
                .path(path)
                .build();
        try {
            root = yaml.load();
        } catch (ConfigurateException e) {
            throw new RuntimeException("Failed to load file", e);
        }
    }

}

File: src/main/kotlin/dev/tuhkanens/comfortlib/api/implementation/ConfigImpl.kt
package dev.tuhkanens.comfortlib.api.implementation

import dev.tuhkanens.comfortlib.api.ConfigAPI
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.yaml.YamlConfigurationLoader
import java.nio.file.Files
import java.nio.file.Path

class ConfigImpl : ConfigAPI {

    private lateinit var root: ConfigurationNode
    private lateinit var yaml: YamlConfigurationLoader
    private lateinit var path: Path

    companion object {
        private const val FILE_NAME = "config.yml"
    }

    override fun create(clazz: Class<*>, directory: Path) {
        path = directory.resolve(FILE_NAME)

        Files.createDirectories(directory)

        if (!Files.exists(path)) {
            val inputStream = clazz.classLoader.getResourceAsStream(FILE_NAME)
                ?: throw IllegalStateException("Default config.yml not found in core resources!")
            inputStream.use { stream ->
                Files.copy(stream, path)
            }
        }

        load()
    }

    override fun getNode(): ConfigurationNode = root
    override fun reload() = load()

    private fun load() {
        yaml = YamlConfigurationLoader.builder()
            .path(path)
            .build()
        root = yaml.load()
    }

}